<div style="background-color:#F3F2EE">
    <br /><br />
        <p style="text-align: center;">
            <font size="6" color='#0A1781'>
                <strong>
                    Spotify Global Music Intelligence - Data Science Projects
                </strong>
              </font>
        </p>
        <p style="text-align: center;">
            <font size="6" color='#C58A1E'>
                <strong>
                    Data Understanding - Entendimento inicial dos dados,<br />estrutura dos arquivos e viabilidade analítica
                </strong>
            </font>
        </p>
        <p style="text-align: center;">
            <font size="5" color='#C58A1E'>
                <strong>
                    Apresentar a finalidade do Notebook 01 e posicionar a etapa de Data Understanding<br />como uma fase essencial para validar a viabilidade analítica do projeto.
                </strong>
            </font>
        </p>
    <br />
</div>

<div style="background-color:#F3F2EE">
    <p style="text-align: right;">
      <font size="4" color='#444444'>
            Roberto SSoares - LfLngLrnng
      </font>
    </p>
    <p style="text-align: right;"><font size="2" color='#444444'>
        <a href="https://www.linkedin.com/in/roberto-dos-santos-soares/">in/roberto-dos-santos-soares</a><br /><a href="https://roberto-ssoares.github.io/meu-portfolio/">Portifólio: roberto-ssoares</a>
    </p>
    <p style="text-align: right;">
        <font size="4" color='#444444'>
            Data Science = { "+"; "Faturamento", "-"; "Custo", ">"; "Qualidade de vida" }
        </font>
        <br />
        <font size="2" color='#918e8e'>"Mestre Bruno Jardim"
        </font>
    </p>        
    <p style="text-align: right;">        
        <font size="2" color='#918e8e'>           
        </font>
    </p>
</div>

<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 Abertura executiva</strong></font>

<font size="2" color='#66666'>

>- Este notebook tem como objetivo iniciar a fase de **Data Understanding** do projeto **Spotify Global Music Intelligence**.

>- Após a definição do problema de negócio, objetivos, stakeholders, perguntas analíticas, hipóteses e critérios de sucesso no Notebook 00,
    - esta etapa avança para a investigação prática dos dados disponíveis.

>- *O foco principal será compreender a estrutura dos arquivos originais, avaliar a granularidade dos dados, identificar colunas disponíveis,*
    - *verificar tipos de variáveis, analisar primeiras evidências de qualidade e avaliar se os datasets podem ser integrados para as próximas fases do projeto.*

---

</font></div>

<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 1. Papel deste notebook no projeto</strong></font>

<font size="2" color='#66666'>

>- O **Notebook 01 — Data Understanding** funciona como uma ponte entre o entendimento de negócio e a preparação dos dados.

>- Antes de aplicar limpeza, transformação, visualizações ou modelagem, é necessário responder perguntas básicas, porém fundamentais:
    - *quais arquivos estão disponíveis;*
    - *quantas linhas e colunas existem em cada dataset;*
    - *quais variáveis estão presentes;*
    - *os arquivos possuem estrutura compatível;*
    - *existem colunas comuns entre as bases;*
    - *existem colunas exclusivas em cada arquivo;*
    - *os tipos de dados fazem sentido;*
    - *há sinais iniciais de valores ausentes;*
    - *há possíveis registros duplicados;*
    - *existe cobertura temporal coerente com o período 2009–2025;*
    - *os dados sustentam as perguntas analíticas formuladas no Notebook 00.*

>- **Ações realizadas nesta célula**
    - Nesta etapa, será definida a função deste notebook dentro do projeto:
        - *iniciar a análise técnica dos arquivos originais;*
        - *validar a existência dos datasets na camada Raw;*
        - *preparar o ambiente de leitura;*
        - *organizar os primeiros diagnósticos;*
        - *documentar decisões iniciais sobre integração e continuidade da análise.*

>- **Justificativa técnica**
    - A fase de entendimento dos dados reduz o risco de avançar para análises exploratórias com bases mal compreendidas, colunas inconsistentes, tipos incorretos, duplicidades ou problemas de granularidade.

>- No contexto deste projeto, essa etapa é especialmente importante porque existem dois arquivos principais:
    - `spotify_data_clean.csv`
    - `track_data_final.csv`

>- Antes de unificar ou comparar essas bases, será necessário compreender se elas possuem estrutura compatível e se representam o mesmo nível de informação analítica.

>- **Resultados esperados**
    - Ao final deste notebook, espera-se obter um diagnóstico inicial contendo:
        - *arquivos localizados corretamente;*
        - *dimensões dos datasets;*
        - *primeiras amostras dos dados;*
        - *lista de colunas disponíveis;*
        - *comparação entre estruturas;*
        - *tipos de dados identificados;*
        - *primeiros sinais de valores ausentes;*
        - *primeiros sinais de duplicidade;*
        - *avaliação preliminar da cobertura temporal;*
        - *recomendação sobre integração ou análise separada dos arquivos.*

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Configuração inicial do ambiente</strong></font>

<font size="2" color='#66666'></font></div>

In [1]:
from pathlib import Path

import numpy as np
import pandas as pd

print("Bibliotecas importadas com sucesso.")

Bibliotecas importadas com sucesso.


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Configurações de exibição do Pandas</strong></font>

<font size="2" color='#66666'></font></div>

In [2]:
# Configurações gerais
pd.set_option("display.max_columns", 200)
pd.set_option("display.max_rows", 200)
pd.set_option("display.max_colwidth", 250)
pd.set_option("display.float_format", "{:,.2f}".format)

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Diretório e Pastas</strong></font>

<font size="2" color='#66666'></font></div>

In [36]:
# Identifica a raiz do projeto
PROJECT_ROOT = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()

# Diretórios principais
DATA_DIR = PROJECT_ROOT / "data"
RAW_DIR = DATA_DIR / "00_raw"
BRONZE_DIR = DATA_DIR / "01_bronze"
SILVER_DIR = DATA_DIR / "02_silver"
GOLD_DIR = DATA_DIR / "03_gold"

REPORTS_DIR = PROJECT_ROOT / "reports"
FIGURES_DIR = REPORTS_DIR / "figures"
HTML_DIR = REPORTS_DIR / "html"

# Arquivos esperados na camada Raw
SPOTIFY_MODERN_FILE = RAW_DIR / "spotify_data_clean.csv"
#SPOTIFY_TRACKS_FILE = RAW_DIR / "track_data_final.csv"
SPOTIFY_TRACKS_FILE = RAW_DIR / "spotify-tracks-dataset.csv"


PROJECT_ROOT

WindowsPath('D:/_DS-Projects/Data-Science/spotify-global-music-intelligence')

<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 2. Validação dos arquivos na camada Raw</strong></font>

<font size="2" color='#66666'>

>- Validar se os arquivos originais do projeto estão disponíveis no diretório `data/00_raw`.

>- **Ações realizadas**
    - Nesta etapa serão verificados:
        - *existência do arquivo `spotify_data_clean.csv`;*
        - *existência do arquivo `track_data_final.csv`;*
        - *caminho completo de cada arquivo;*
        - *status inicial de disponibilidade dos dados.*

>- **Justificativa técnica**
    - Antes de carregar os dados em memória, é necessário garantir que os arquivos esperados estejam corretamente posicionados na camada Raw.
    - Essa validação evita erros posteriores de leitura e ajuda a documentar o ponto de partida do pipeline analítico.

>- **Resultado esperado**
    - Ao final desta célula, espera-se confirmar se os dois arquivos principais do projeto estão disponíveis para leitura.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Validação da existência dos arquivos na camada Raw</strong></font>

<font size="2" color='#66666'></font></div>

In [37]:
raw_files_check = pd.DataFrame(
    {
        "arquivo_esperado": [
            "spotify_data_clean.csv",
            "spotify-tracks-dataset.csv",
        ],
        "caminho": [
            
            SPOTIFY_MODERN_FILE,
            SPOTIFY_TRACKS_FILE,
        ],
        "existe": [
            SPOTIFY_MODERN_FILE.exists(),
            SPOTIFY_TRACKS_FILE.exists(),
        ],
    }
)

raw_files_check

,arquivo_esperado,caminho,existe
0,spotify_data_clean.csv,D:\_DS-Projects\Data-Science\spotify-global-music-intelligence\data\00_raw\spotify_data_clean.csv,True
1,spotify-tracks-dataset.csv,D:\_DS-Projects\Data-Science\spotify-global-music-intelligence\data\00_raw\spotify-tracks-dataset.csv,True


<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 3. Carregamento inicial dos datasets</strong></font>

<font size="2" color='#66666'>

>- Carregar os arquivos CSV originais disponíveis na camada `data/00_raw` para iniciar o entendimento técnico dos dados.

>- **Ações realizadas**
    - Nesta etapa serão carregados os seguintes arquivos:
        - `spotify_data_clean.csv`
        - `track_data_final.csv`
    - Cada arquivo será lido como um DataFrame Pandas independente, preservando inicialmente sua estrutura original.

>- **Justificativa técnica**
    - Antes de realizar qualquer limpeza, padronização ou integração, é necessário carregar os dados em seu formato bruto e observar como eles estão estruturados.
    - Essa prática preserva a rastreabilidade entre a camada Raw e as próximas etapas do projeto.

>- **Resultado esperado**
    - Ao final desta célula, espera-se ter dois DataFrames carregados:
        - `df_spotify_modern`
        - `df_spotify_tracks`
    - Esses DataFrames serão usados nas próximas células para análise de dimensões, colunas, tipos de dados e primeiras amostras.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Carregamento inicial dos arquivos CSV</strong></font>

<font size="2" color='#66666'></font></div>

In [29]:
def read_csv_with_fallback(file_path: Path) -> pd.DataFrame:
    """
    Lê um arquivo CSV usando tentativas de encoding e separador.

    A função tenta combinações comuns para reduzir erros de leitura,
    especialmente em datasets públicos baixados de plataformas externas.
    """
    attempts = [
        {"encoding": "utf-8", "sep": ","},
        {"encoding": "utf-8-sig", "sep": ","},
        {"encoding": "latin1", "sep": ","},
        {"encoding": "utf-8", "sep": ";"},
        {"encoding": "utf-8-sig", "sep": ";"},
        {"encoding": "latin1", "sep": ";"},
    ]

    last_error = None

    for params in attempts:
        try:
            return pd.read_csv(file_path, low_memory=False, **params)
        except Exception as error:
            last_error = error

    raise RuntimeError(
        f"Não foi possível ler o arquivo {file_path}. "
        f"Último erro encontrado: {last_error}"
    )
    

In [38]:
df_spotify_modern = read_csv_with_fallback(SPOTIFY_MODERN_FILE)
df_spotify_tracks = read_csv_with_fallback(SPOTIFY_TRACKS_FILE)

print("Datasets carregados com sucesso.")

Datasets carregados com sucesso.


<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 4. Diagnóstico inicial de volume</strong></font>

<font size="2" color='#66666'>

>- Verificar o tamanho inicial de cada dataset carregado.

>- **Ações realizadas**
    - Nesta etapa serão analisadas:
        - *quantidade de linhas;*
        - *quantidade de colunas;*
        - *quantidade total de células;*
        - *nome lógico de cada dataset.*

>- **Justificativa técnica**
    - O volume inicial dos dados ajuda a entender a escala do projeto, avaliar o esforço de preparação e identificar possíveis diferenças relevantes entre os arquivos.
    - Essa análise também permite verificar se os arquivos foram carregados corretamente.

>- **Resultado esperado**
    - Ao final desta célula, espera-se obter uma tabela comparativa com as dimensões iniciais dos dois datasets.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Diagnóstico inicial de volume dos datasets</strong></font>

<font size="2" color='#66666'></font></div>

In [39]:
volume_report = pd.DataFrame(
    {
        "dataset": [
            "spotify_data_clean",
            "track_data_final",
        ],
        "rows": [
            df_spotify_modern.shape[0],
            df_spotify_tracks.shape[0],
        ],
        "columns": [
            df_spotify_modern.shape[1],
            df_spotify_tracks.shape[1],
        ],
    }
)

volume_report["total_cells"] = volume_report["rows"] * volume_report["columns"]

volume_report

,dataset,rows,columns,total_cells
0,spotify_data_clean,8582,15,128730
1,track_data_final,114000,21,2394000


<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 5. Primeiras amostras dos dados</strong></font>

<font size="2" color='#66666'>

>- Visualizar os primeiros registros de cada dataset para compreender sua estrutura inicial.

>- **Ações realizadas**
    - Nesta etapa serão exibidas amostras dos dois DataFrames:
        - *primeiras linhas de `df_spotify_modern`;*
        - *primeiras linhas de `df_spotify_tracks`.*

>- **Justificativa técnica**
    - A visualização inicial das linhas permite identificar rapidamente nomes de colunas, formatos de valores,
        - possíveis problemas de separador, presença de campos textuais e variáveis candidatas para análise.

>- **Resultado esperado**
    - Ao final desta célula, espera-se obter uma primeira leitura visual da estrutura dos dados.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Primeiras linhas dos datasets</strong></font>

<font size="2" color='#66666'></font></div>

In [40]:
df_spotify_modern.head()

,track_id,track_name,track_number,track_popularity,explicit,artist_name,artist_popularity,artist_followers,artist_genres,album_id,album_name,album_release_date,album_total_tracks,album_type,track_duration_min
0,3EJS5LyekDim1Tf5rBFmZl,Trippy Mane (ft. Project Pat),4,0,True,Diplo,77,2812821,moombahton,5QRFnGnBeMGePBKF2xTz5z,"d00mscrvll, Vol. 1",2025-10-31,9,album,1.55
1,1oQW6G2ZiwMuHqlPpP27DB,OMG!,1,0,True,Yelawolf,64,2363438,"country hip hop, southern hip hop",4SUmmwnv0xTjRcLdjczGg2,OMG!,2025-10-31,1,single,3.07
2,7mdkjzoIYlf1rx9EtBpGmU,Hard 2 Find,1,4,True,Riff Raff,48,193302,NaN,3E3zEAL8gUYWaLYB9L7gbp,Hard 2 Find,2025-10-31,1,single,2.55
3,67rW0Zl7oB3qEpD5YWWE5w,Still Get Like That (ft. Project Pat & Starrah),8,30,True,Diplo,77,2813710,moombahton,5QRFnGnBeMGePBKF2xTz5z,"d00mscrvll, Vol. 1",2025-10-31,9,album,1.69
4,15xptTfRBrjsppW0INUZjf,ride me like a harley,2,0,True,Rumelis,48,8682,dark r&b,06FDIpSHYmZAZoyuYtc7kd,come closer / ride me like a harley,2025-10-30,2,single,2.39


In [41]:
df_spotify_tracks.head()

,Unnamed: 0,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
0,0,5SuOikwiRyPMVoIQDJUgSV,Gen Hoshino,Comedy,Comedy,73,230666,False,0.68,0.46,1,-6.75,0,0.14,0.03,0.00,0.36,0.71,87.92,4,acoustic
1,1,4qPNDBW1i3p13qLCt0Ki3A,Ben Woodward,Ghost (Acoustic),Ghost - Acoustic,55,149610,False,0.42,0.17,1,-17.23,1,0.08,0.92,0.00,0.10,0.27,77.49,4,acoustic
2,2,1iJBSr7s7jYXzM8EGcbK5b,Ingrid Michaelson;ZAYN,To Begin Again,To Begin Again,57,210826,False,0.44,0.36,0,-9.73,1,0.06,0.21,0.00,0.12,0.12,76.33,4,acoustic
3,3,6lfxq3CG4xtTiEg7opyCyx,Kina Grannis,Crazy Rich Asians (Original Motion Picture Soundtrack),Can't Help Falling In Love,71,201933,False,0.27,0.06,0,-18.52,1,0.04,0.91,0.00,0.13,0.14,181.74,3,acoustic
4,4,5vjLSffimiIP26QG5WcN2K,Chord Overstreet,Hold On,Hold On,82,198853,False,0.62,0.44,2,-9.68,1,0.05,0.47,0.00,0.08,0.17,119.95,4,acoustic


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Comparação inicial das colunas dos dois datasets</strong></font>

<font size="2" color='#66666'></font></div>

In [47]:
modern_columns = set(df_spotify_modern.columns)
tracks_columns = set(df_spotify_tracks.columns)

common_columns = sorted(modern_columns.intersection(tracks_columns))
only_modern_columns = sorted(modern_columns.difference(tracks_columns))
only_tracks_columns = sorted(tracks_columns.difference(modern_columns))

schema_comparison = pd.DataFrame(
    {
        "categoria": [
            "colunas_em_comum",
            "somente_spotify_data_clean",
            "somente_spotify_tracks_dataset",
        ],
        "quantidade": [
            len(common_columns),
            len(only_modern_columns),
            len(only_tracks_columns),
        ],
        "colunas": [
            common_columns,
            only_modern_columns,
            only_tracks_columns,
        ],
    }
)

schema_comparison

,categoria,quantidade,colunas
0,colunas_em_comum,4,"[album_name, explicit, track_id, track_name]"
1,somente_spotify_data_clean,11,"[album_id, album_release_date, album_total_tracks, album_type, artist_followers, artist_genres, artist_name, artist_popularity, track_duration_min, track_number, track_popularity]"
2,somente_spotify_tracks_dataset,17,"[Unnamed: 0, acousticness, artists, danceability, duration_ms, energy, instrumentalness, key, liveness, loudness, mode, popularity, speechiness, tempo, time_signature, track_genre, valence]"


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Resumo das colunas em comum e exclusivas</strong></font>

<font size="2" color='#66666'></font></div>

In [46]:
print("Colunas em comum:")
print(common_columns)

print("\nColunas somente em spotify_data_clean:")
print(only_modern_columns)

print("\nColunas somente em spotify_tracks_dataset:")
print(only_tracks_columns)
print(" ")
print("==="*30)
print(" ")
print(df_spotify_modern.info())
print(" ")
print("==="*30)
print(" ")
print(df_spotify_tracks.info())

Colunas em comum:
['album_name', 'explicit', 'track_id', 'track_name']

Colunas somente em spotify_data_clean:
['album_id', 'album_release_date', 'album_total_tracks', 'album_type', 'artist_followers', 'artist_genres', 'artist_name', 'artist_popularity', 'track_duration_min', 'track_number', 'track_popularity']

Colunas somente em spotify_tracks_dataset:
['Unnamed: 0', 'acousticness', 'artists', 'danceability', 'duration_ms', 'energy', 'instrumentalness', 'key', 'liveness', 'loudness', 'mode', 'popularity', 'speechiness', 'tempo', 'time_signature', 'track_genre', 'valence']
 
 
<class 'pandas.DataFrame'>
RangeIndex: 8582 entries, 0 to 8581
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   track_id            8582 non-null   str    
 1   track_name          8582 non-null   str    
 2   track_number        8582 non-null   int64  
 3   track_popularity    8582 non-null   int64  
 4   explicit            

<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 6. Interpretação da comparação inicial de schema</strong></font>

<font size="2" color='#66666'>

>- Registrar a interpretação técnica inicial sobre a compatibilidade e complementaridade entre os datasets selecionados para continuidade do projeto:
    - `spotify_data_clean.csv`
    - `spotify-tracks-dataset.csv`
    - Após a análise preliminar, foi tomada a decisão de seguir com essas duas fontes principais, deixando o arquivo `track_data_final.csv` fora do escopo imediato desta etapa.

>- **Decisão metodológica tomada**
    - A partir deste ponto, o projeto utilizará como fontes principais:

>| Papel no projeto | Arquivo | Descrição |
>|---|---|---|
>| Fonte temporal e contextual | `spotify_data_clean.csv` | Dataset com informações de faixa, artista, álbum, popularidade, data de lançamento e metadados contextuais |
>| Fonte musical e comportamental | `spotify-tracks-dataset.csv` | Dataset mais amplo, com atributos musicais, gênero, características sonoras e popularidade |

>- A decisão foi tomada porque os dois datasets possuem estruturas diferentes e complementares, o que cria uma oportunidade mais rica para análise exploratória, engenharia de atributos e futura construção de uma ABT.

>- **Resultado observado na comparação de colunas**
    - A comparação de schema indica que os datasets não são apenas variações um do outro.
    - O arquivo `spotify_data_clean.csv` possui colunas mais voltadas a contexto de artista, álbum, lançamento e popularidade no recorte original do projeto.
    - Já o arquivo `spotify-tracks-dataset.csv` possui variáveis musicais mais detalhadas, especialmente atributos sonoros e gênero.

>- **Colunas exclusivas de `spotify_data_clean.csv`**

>| Grupo | Colunas |
>|---|---|
>| Álbum | `album_id`, `album_release_date`, `album_total_tracks`, `album_type` |
>| Artista | `artist_name`, `artist_popularity`, `artist_followers`, `artist_genres` |
>| Faixa | `track_number`, `track_popularity`, `track_duration_min` |

>- **Colunas exclusivas de `spotify-tracks-dataset.csv`**

>| Grupo | Colunas |
>|---|---|
>| Atributos musicais | `danceability`, `energy`, `loudness`, `speechiness`, `acousticness`, `instrumentalness`, `liveness`, `valence`, `tempo` |
>| Estrutura musical | `key`, `mode`, `time_signature`, `duration_ms` |
>| Gênero | `track_genre` |
>| Popularidade | `popularity` |
>| Artistas | `artists` |
>| Controle técnico | `Unnamed: 0` |

>- **Interpretação técnica**
    - Os dois datasets parecem representar a mesma entidade analítica principal: a faixa musical.
    - No entanto, eles possuem perspectivas diferentes sobre essa entidade.

>| Dataset | Perspectiva predominante | Valor analítico |
>|---|---|---|
>| `spotify_data_clean.csv` | Contexto de faixa, artista, álbum e lançamento | Apoia análise temporal, análise por artista e contexto de popularidade |
>| `spotify-tracks-dataset.csv` | Características musicais e gênero | Apoia análise sonora, segmentação, clustering e modelagem futura |

>- Essa diferença é positiva para o projeto, pois permite construir uma análise mais rica do que seria possível com apenas uma das fontes.

>- **Complementaridade analítica**
    - A principal vantagem de utilizar os dois datasets é combinar duas dimensões importantes:
        1. **Dimensão contextual**
           - artista;
           - popularidade do artista;
           - seguidores;
           - álbum;
           - data de lançamento;
           - tipo de álbum;
           - número da faixa;
           - popularidade da faixa.
        2. **Dimensão musical**
           - dançabilidade;
           - energia;
           - acústica;
           - instrumentalidade;
           - vivacidade;
           - valência;
           - tempo;
           - intensidade sonora;
           - gênero musical;
           - assinatura de tempo;
           - tonalidade e modo.

>- Essa combinação pode apoiar uma futura ABT com variáveis explicativas mais completas.

>- **Diferenças semânticas importantes**
    - Algumas colunas possuem nomes diferentes, mas representam conceitos próximos.

>| Conceito | `spotify_data_clean.csv` | `spotify-tracks-dataset.csv` | Ação necessária |
>|---|---|---|---|
>| Popularidade da faixa | `track_popularity` | `popularity` | Padronizar nome e avaliar escala |
>| Duração da faixa | `track_duration_min` | `duration_ms` | Converter para uma unidade comum |
>| Artista | `artist_name` | `artists` | Avaliar formato e padronizar |
>| Gênero | `artist_genres` | `track_genre` | Tratar como perspectivas diferentes |
>| Nome da faixa | `track_name` | `track_name` | Usar como possível chave auxiliar |
>| Nome do álbum | `album_name` | `album_name` | Usar como possível chave auxiliar |

>- Essas diferenças precisam ser tratadas antes de qualquer integração definitiva.

>- **Risco de integração direta**
    - Apesar da complementaridade, não é recomendável integrar os datasets imediatamente sem validação.
    - Os principais riscos são:
        - `track_id` pode não ter interseção suficiente entre as bases;
        - a mesma música pode aparecer com identificadores diferentes;
        - nomes de artistas podem ter formatos distintos;
        - a coluna `artists` pode conter múltiplos artistas em uma única string;
        - `artist_genres` e `track_genre` podem representar conceitos diferentes;
        - `track_popularity` e `popularity` podem ter sido coletadas em momentos diferentes;
        - duplicidades podem surgir ao unir bases com granularidade semelhante;
        - a coluna `Unnamed: 0` deve ser avaliada e provavelmente removida.

>- **Decisão preliminar**
    - A decisão preliminar é utilizar os dois datasets de forma complementar, mas com cautela.
    - Nesta fase de Data Understanding, os datasets serão analisados separadamente para compreender:
        - estrutura;
        - volume;
        - tipos de dados;
        - valores ausentes;
        - duplicidades;
        - cobertura temporal;
        - colunas em comum;
        - colunas exclusivas;
        - possibilidade real de integração.

>- A integração será considerada apenas depois de avaliar a interseção por `track_id` e, se necessário, por chaves auxiliares como `track_name`, `album_name` e artista.

>- **Estratégia recomendada para a ABT**
    - A estratégia recomendada para uma futura ABT é:
        1. manter `spotify_data_clean.csv` como fonte contextual;
        2. manter `spotify-tracks-dataset.csv` como fonte de atributos musicais;
        3. padronizar nomes de colunas;
        4. converter duração para uma unidade comum;
        5. harmonizar popularidade como variável analítica;
        6. avaliar interseção por `track_id`;
        7. avaliar chaves alternativas com cautela;
        8. criar uma base enriquecida apenas com registros confiáveis;
        9. preservar uma coluna de origem dos dados;
        10. documentar perdas, duplicidades e limitações da integração.

>- **Papel de cada dataset no projeto**

>| Dataset | Uso recomendado nesta fase |
>|---|---|
>| `spotify_data_clean.csv` | Análise temporal, análise de artistas, álbuns, popularidade e contexto de lançamento |
>| `spotify-tracks-dataset.csv` | Análise de atributos musicais, gêneros, segmentação, correlação e preparação para modelagem |
>| Base ABT futura | Combinação controlada das duas fontes, caso a integração seja tecnicamente confiável |

>- **Resultado esperado**
    - Esta interpretação orienta a próxima etapa do Data Understanding.
    - A partir deste ponto, o projeto deixa de tratar os arquivos como bases redundantes e passa a tratá-los como fontes complementares.
    - Essa decisão fortalece o projeto porque permite construir uma análise mais completa, conectando:
        - contexto musical;
        - dados de artista;
        - dados de álbum;
        - popularidade;
        - atributos sonoros;
        - gênero musical;
        - potencial de segmentação;
        - preparação para machine learning.

>- A próxima etapa será avaliar a interseção entre os datasets por `track_id` e diagnosticar se existe viabilidade real para enriquecimento entre as fontes.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 7. Decisão metodológica: foco na base principal</strong></font>

<font size="2" color='#66666'>

>- Registrar a decisão metodológica de seguir, nesta etapa do projeto, com o dataset `spotify_data_clean.csv` como fonte principal da análise exploratória.

>- **Contexto da decisão**
    - Durante o Data Understanding, foram avaliadas outras fontes possíveis, incluindo datasets complementares com atributos musicais mais detalhados.
    - No entanto, para manter o projeto alinhado ao planejamento inicial do Cap01 — EDA, a decisão tomada foi concentrar a primeira fase analítica na base `spotify_data_clean.csv`.
    - Essa decisão evita dispersão, reduz complexidade prematura e permite construir uma EDA mais controlada, clara e bem documentada.

>- **Papel da base `spotify_data_clean.csv`**
    - A base principal contém informações relevantes para uma análise exploratória inicial, incluindo:
        - identificação da faixa;
        - nome da música;
        - popularidade da faixa;
        - artista;
        - popularidade do artista;
        - seguidores do artista;
        - gêneros do artista;
        - álbum;
        - data de lançamento;
        - tipo de álbum;
        - duração da faixa.
    - Essas variáveis são suficientes para investigar popularidade, artistas, álbuns, lançamentos, gêneros e evolução temporal.

>- **Fonte complementar futura**
    - O arquivo `spotify-tracks-dataset.csv` será preservado para uma etapa futura.
    - Seu uso potencial inclui:
        - enriquecimento da base principal;
        - criação de uma ABT;
        - análise de atributos musicais;
        - segmentação;
        - clusterização;
        - modelagem preditiva;
        - classificação de potencial de popularidade.

>- **Decisão final desta fase**
    - Nesta fase do projeto, a EDA será conduzida exclusivamente sobre:
    - `spotify_data_clean.csv`
    - A base complementar será retomada somente após a conclusão da análise exploratória principal.

>- **Resultado esperado**
    - A partir desta decisão, o Notebook 01 seguirá com o entendimento aprofundado da base principal, avaliando estrutura, tipos, nulos,
        - duplicidades, cardinalidade, cobertura temporal e primeiras estatísticas descritivas.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Perfil técnico inicial da base principal</strong></font>

<font size="2" color='#66666'></font></div>

In [48]:
df_main = df_spotify_modern.copy()

profile_columns = pd.DataFrame(
    {
        "column": df_main.columns,
        "dtype": [df_main[col].dtype for col in df_main.columns],
        "non_null_count": [df_main[col].notna().sum() for col in df_main.columns],
        "null_count": [df_main[col].isna().sum() for col in df_main.columns],
        "null_pct": [df_main[col].isna().mean() * 100 for col in df_main.columns],
        "unique_count": [df_main[col].nunique(dropna=True) for col in df_main.columns],
        "unique_pct": [
            df_main[col].nunique(dropna=True) / len(df_main) * 100
            for col in df_main.columns
        ],
    }
)

profile_columns.sort_values(["null_pct", "unique_count"], ascending=[False, False])

,column,dtype,non_null_count,null_count,null_pct,unique_count,unique_pct
8,artist_genres,str,5221,3361,39.16,661,7.70
5,artist_name,str,8579,3,0.03,2547,29.68
0,track_id,str,8582,0,0.00,8582,100.00
1,track_name,str,8582,0,0.00,7462,86.95
9,album_id,str,8582,0,0.00,5205,60.65
10,album_name,str,8582,0,0.00,4870,56.75
7,artist_followers,int64,8582,0,0.00,3740,43.58
11,album_release_date,str,8582,0,0.00,2384,27.78
14,track_duration_min,float64,8582,0,0.00,647,7.54
3,track_popularity,int64,8582,0,0.00,98,1.14


<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 8. Interpretação do perfil técnico das colunas</strong></font>

<font size="2" color='#66666'>

>- Registrar a interpretação inicial sobre tipos de dados, completude, cardinalidade e qualidade das variáveis da base `spotify_data_clean.csv`.

>- **Resultado observado**
    - A base principal possui 8.582 registros e 15 colunas.
    - De forma geral, a estrutura dos dados é adequada para continuidade da análise exploratória. A variável `track_id` apresenta 8.582 valores únicos,
        - indicando que cada linha possui um identificador técnico único de faixa musical.
    - Também não foram identificados valores ausentes nas principais variáveis de análise, como:
        - `track_name`;
        - `track_popularity`;
        - `artist_popularity`;
        - `artist_followers`;
        - `album_name`;
        - `album_release_date`;
        - `track_duration_min`;
        - `explicit`;
        - `album_type`.
    - Esse resultado é positivo, pois indica que a base possui boa completude nas variáveis centrais do projeto.

>- **Pontos positivos identificados**

>| Aspecto | Interpretação |
>|---|---|
>| `track_id` 100% único | Indica boa chave técnica para análise por faixa |
>| `track_popularity` sem nulos | Permite usar popularidade como eixo central da EDA |
>| `artist_popularity` sem nulos | Permite comparar força do artista com desempenho da faixa |
>| `artist_followers` sem nulos | Permite avaliar relação entre base de fãs e popularidade |
>| `album_release_date` sem nulos | Permite análise temporal |
>| `track_duration_min` sem nulos | Permite análise de duração das faixas |
>| `explicit` sem nulos | Permite análise entre faixas explícitas e não explícitas |

>- **Pontos de atenção identificados**
    - O principal ponto de atenção está na coluna `artist_genres`, que apresenta 3.361 registros ausentes, correspondendo a aproximadamente 39,16% da base.
    - Esse percentual é relevante e exige cuidado nas análises por gênero.
    - Também foram identificados 3 registros sem `artist_name`, um volume muito baixo, mas que será investigado antes de qualquer tratamento.

>- **Decisão metodológica**
    - Nesta etapa, nenhuma linha será removida automaticamente.
    - As decisões preliminares são:
        - manter todos os registros nesta fase de Data Understanding;
        - investigar os 3 registros sem `artist_name`;
        - tratar `artist_genres` ausente como informação não disponível;
        - converter `album_release_date` para data;
        - criar variáveis temporais auxiliares, como `release_year`;
        - documentar limitações antes de análises por gênero.

>- **Resultado esperado**
    - A base `spotify_data_clean.csv` será mantida como fonte principal da EDA.
    - Os próximos passos serão investigar valores ausentes, duplicidades textuais e cobertura temporal antes de avançar para a preparação formal dos dados no Notebook 02.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Investigação dos registros sem artist_name</strong></font>

<font size="2" color='#66666'></font></div>

In [56]:
missing_artist_name = df_main[df_main["artist_name"].isna()]

missing_artist_name[
    [
        "track_id",
        "track_name",
        "track_popularity",
        "artist_name",
        "artist_popularity",
        "artist_followers",
        "artist_genres",
        "album_name",
        "album_release_date",
        "album_type",
    ]
]


,track_id,track_name,track_popularity,artist_name,artist_popularity,artist_followers,artist_genres,album_name,album_release_date,album_type
1807,2NbsBSshscKIPRG3kwBbfo,Never Felt So Alone,68,NaN,0,0,NaN,Never Felt So Alone,2023-04-07,single
6517,3QhfFRPkhPCR1RMJWV1gde,Radio,65,NaN,0,0,NaN,Born To Die – Paradise Edition (Special Version),2012-01-01,album
6679,5EhZC8GFOVXrpiUW22jncE,Urban Twilight,30,NaN,0,0,NaN,Darkbloom,2011-04-18,single


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Diagnóstico dos registros sem artist_genres</strong></font>

<font size="2" color='#66666'></font></div>

In [57]:
genre_missing_summary = pd.DataFrame(
    {
        "metric": [
            "total_rows",
            "rows_with_artist_genres",
            "rows_without_artist_genres",
            "artist_genres_missing_pct",
            "unique_artists_without_genres",
            "avg_track_popularity_without_genres",
            "avg_track_popularity_with_genres",
        ],
        "value": [
            len(df_main),
            df_main["artist_genres"].notna().sum(),
            df_main["artist_genres"].isna().sum(),
            round(df_main["artist_genres"].isna().mean() * 100, 2),
            df_main.loc[df_main["artist_genres"].isna(), "artist_name"].nunique(dropna=True),
            round(df_main.loc[df_main["artist_genres"].isna(), "track_popularity"].mean(), 2),
            round(df_main.loc[df_main["artist_genres"].notna(), "track_popularity"].mean(), 2),
        ],
    }
)

genre_missing_summary

,metric,value
0,total_rows,"8,582.00"
1,rows_with_artist_genres,"5,221.00"
2,rows_without_artist_genres,"3,361.00"
3,artist_genres_missing_pct,39.16
4,unique_artists_without_genres,"1,336.00"
5,avg_track_popularity_without_genres,50.49
6,avg_track_popularity_with_genres,53.56


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Amostra de registros sem artist_genres</strong></font>

<font size="2" color='#66666'></font></div>

In [58]:
df_main.loc[
    df_main["artist_genres"].isna(),
    [
        "track_id",
        "track_name",
        "artist_name",
        "track_popularity",
        "artist_popularity",
        "artist_followers",
        "album_name",
        "album_release_date",
    ],
].head(20)

,track_id,track_name,artist_name,track_popularity,artist_popularity,artist_followers,album_name,album_release_date
2,7mdkjzoIYlf1rx9EtBpGmU,Hard 2 Find,Riff Raff,4,48,193302,Hard 2 Find,2025-10-31
6,3QoQ3HqXTAjgEl9LbNMbYp,Te Procuro na Cidade,AZERDK,20,30,1657,Te Procuro na Cidade,2025-10-30
8,4pZ949nFW5SurwzE0TSe7I,Cupido Vagabundo,Toni dos Anjos,16,6,475,Cupido Vagabundo,2025-10-30
10,5mfAZwEr4jAVDnSY0HHuyH,Stray,What Matters The Most,30,25,814,Stray,2025-10-25
12,0NZKfcua68wuZePhjp8N2I,Root of all Evil,Daniel Caesar,65,85,11137714,Son Of Spergy,2025-10-24
13,6zFWxD9hkOPAIlgWAGyQGQ,Let’s Go!!,Ovi Jozi,0,17,3567,"High Risk, High Reward",2025-10-24
15,22c6Yop5YVEAhw8UG6O1W1,My Muse,Leon Thomas,67,74,735818,PHOLKS,2025-10-24
16,0SoYN9bRGPj7qV6loQKCTO,A Real Good Bad One,Joe Hermes,31,23,1518,A Real Good Bad One,2025-10-24
18,0XQQpq855N5rnUsFBXo2ml,Bridges,Scarlett Butler,11,3,364,Bridges,2025-10-24
19,0nBPI5KxKOeOtr24sdGpE6,Can't Go Back,LEROCQUE,30,23,13770,Can't Go Back,2025-10-24


<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 9. Diagnóstico inicial de duplicidade</strong></font>

<font size="2" color='#66666'>

>- Avaliar se a base principal possui registros duplicados que possam distorcer análises de popularidade, artistas, álbuns e faixas.

>- **Ações realizadas**
    - Nesta etapa serão verificados:
        - *registros completamente duplicados;*
        - *duplicidade por `track_id`;*
        - *duplicidade por combinação de `track_name`, `artist_name` e `album_name`.*

>- **Justificativa técnica**
    - Duplicidades podem inflar contagens, alterar médias de popularidade e comprometer rankings.
    - Antes de avançar para a EDA, é necessário entender se cada linha representa uma faixa única ou se existem repetições relevantes.

>- **Resultado esperado**
    - Obter um diagnóstico inicial sobre a unicidade dos registros da base principal.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Diagnóstico de duplicidade da base principal</strong></font>

<font size="2" color='#66666'></font></div>

In [49]:
duplicate_summary = pd.DataFrame(
    {
        "metric": [
            "total_rows",
            "unique_track_id",
            "duplicated_full_rows",
            "duplicated_track_id",
            "duplicated_track_artist_album",
        ],
        "value": [
            len(df_main),
            df_main["track_id"].nunique(dropna=True),
            df_main.duplicated().sum(),
            df_main["track_id"].duplicated().sum(),
            df_main.duplicated(
                subset=["track_name", "artist_name", "album_name"]
            ).sum(),
        ],
    }
)

duplicate_summary

,metric,value
0,total_rows,8582
1,unique_track_id,8582
2,duplicated_full_rows,0
3,duplicated_track_id,0
4,duplicated_track_artist_album,117


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Cobertura temporal inicial da base principal</strong></font>

<font size="2" color='#66666'></font></div>

In [50]:
df_main["album_release_date_parsed"] = pd.to_datetime(
    df_main["album_release_date"],
    errors="coerce"
)

df_main["release_year"] = df_main["album_release_date_parsed"].dt.year

temporal_summary = pd.DataFrame(
    {
        "metric": [
            "min_release_date",
            "max_release_date",
            "min_release_year",
            "max_release_year",
            "records_with_invalid_release_date",
            "unique_release_years",
        ],
        "value": [
            df_main["album_release_date_parsed"].min(),
            df_main["album_release_date_parsed"].max(),
            df_main["release_year"].min(),
            df_main["release_year"].max(),
            df_main["album_release_date_parsed"].isna().sum(),
            df_main["release_year"].nunique(dropna=True),
        ],
    }
)

temporal_summary

,metric,value
0,min_release_date,1952-09-12 00:00:00
1,max_release_date,2025-10-31 00:00:00
2,min_release_year,1952
3,max_release_year,2025
4,records_with_invalid_release_date,0
5,unique_release_years,68


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Quantidade de registros por ano de lançamento</strong></font>

<font size="2" color='#66666'></font></div>

In [53]:
tracks_by_year = (
    df_main
    .groupby("release_year", dropna=False)
    .size()
    .reset_index(name="track_count")
    .sort_values("release_year")
)

tracks_by_year

,release_year,track_count
0,1952,1
1,1958,1
2,1959,5
3,1960,1
4,1962,5
5,1963,1
6,1964,1
7,1965,3
8,1966,8
9,1967,11


<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 10. Interpretação inicial de unicidade e cobertura temporal</strong></font>

<font size="2" color='#66666'>

>- **Diagnóstico de duplicidade**
    - A análise inicial de duplicidade da base `spotify_data_clean.csv` indicou que o dataset possui 8.582 registros e 8.582 valores únicos para `track_id`.
    - Esse resultado sugere que cada linha representa uma faixa musical única do ponto de vista do identificador técnico fornecido pela base.
    - Também não foram identificadas linhas completamente duplicadas nem duplicidade direta por `track_id`.
    - No entanto, foram encontrados 117 casos de duplicidade considerando a combinação:
        - `track_name + artist_name + album_name`
    - Esses casos não serão removidos automaticamente nesta etapa, pois podem representar versões diferentes da mesma música,
        - relançamentos, singles, faixas explícitas ou limpas, variações de álbum ou diferenças reais de catálogo.
    - A decisão inicial é manter esses registros e investigá-los antes de qualquer regra de deduplicação.

>- **Diagnóstico temporal**
    - A conversão da coluna `album_release_date` para formato de data foi realizada com sucesso, sem registros inválidos.
    - A base apresenta cobertura temporal entre 1952 e 2025, totalizando 68 anos distintos de lançamento.
    - Embora o projeto tenha sido inicialmente posicionado com foco em 2009–2025, a presença de faixas anteriores a 2009 indica que o dataset contém também músicas de catálogo histórico.
    - Esse achado é relevante porque permite diferenciar:
        - faixas históricas ou clássicas;
        - músicas do período moderno do streaming;
        - lançamentos recentes de 2024 e 2025.

>- **Decisão metodológica**
    - A decisão metodológica nesta etapa é não descartar automaticamente os registros anteriores a 2009.
    - Em vez disso, será criada uma variável de período analítico para classificar as faixas em grupos temporais.
    - Essa abordagem preserva informação histórica, mantém rastreabilidade e permite comparar o comportamento de músicas antigas e recentes durante a EDA.

>- **Implicações para as próximas etapas**
    - Os próximos passos serão:
        1. investigar os 117 casos de possível duplicidade textual;
        2. criar uma variável `periodo_analitico`;
        3. avaliar a concentração de registros por período;
        4. decidir se a EDA principal usará todos os registros ou apenas o recorte 2009–2025;
        5. documentar essa decisão antes do Notebook 02.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Criação de período analítico para análise temporal</strong></font>

<font size="2" color='#66666'></font></div>

In [54]:
def classify_release_period(year: float) -> str:
    """
    Classifica o ano de lançamento em períodos analíticos.
    """
    if pd.isna(year):
        return "unknown"

    year = int(year)

    if year < 2009:
        return "catalogo_historico_pre_2009"
    if 2009 <= year <= 2015:
        return "streaming_early_2009_2015"
    if 2016 <= year <= 2020:
        return "streaming_growth_2016_2020"
    if 2021 <= year <= 2023:
        return "streaming_mature_2021_2023"
    if 2024 <= year <= 2025:
        return "recent_releases_2024_2025"

    return "out_of_scope"


df_main["periodo_analitico"] = df_main["release_year"].apply(classify_release_period)

period_summary = (
    df_main
    .groupby("periodo_analitico", dropna=False)
    .agg(
        track_count=("track_id", "count"),
        min_year=("release_year", "min"),
        max_year=("release_year", "max"),
        avg_track_popularity=("track_popularity", "mean"),
        median_track_popularity=("track_popularity", "median"),
    )
    .reset_index()
    .sort_values("min_year")
)

period_summary

,periodo_analitico,track_count,min_year,max_year,avg_track_popularity,median_track_popularity
0,catalogo_historico_pre_2009,1394,1952,2008,54.07,59.00
2,streaming_early_2009_2015,2009,2009,2015,51.44,56.00
3,streaming_growth_2016_2020,2332,2016,2020,52.22,59.00
4,streaming_mature_2021_2023,1434,2021,2023,51.90,59.00
1,recent_releases_2024_2025,1413,2024,2025,52.66,54.00


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Investigação de possíveis duplicidades textuais na base principal</strong></font>

<font size="2" color='#66666'></font></div>

In [65]:
possible_text_duplicates = (
    df_main[
        df_main.duplicated(
            subset=["track_name", "artist_name", "album_name"],
            keep=False,
        )
    ]
    .sort_values(["artist_name", "album_name", "track_name", "track_number"])
)

possible_text_duplicates[
    [
        "track_id",
        "track_name",
        "artist_name",
        "album_name",
        "album_release_date",
        "track_number",
        "track_popularity",
        "explicit",
        "track_duration_min",
    ]
].head(50)

,track_id,track_name,artist_name,album_name,album_release_date,track_number,track_popularity,explicit,track_duration_min
7734,3NLrRZoMF0Lx6zTlYqeIo4,Here Without You,3 Doors Down,Away From The Sun,2002-11-12,6,79,False,3.97
7804,5IXk34xSm6rxNgXkCveC3r,Here Without You,3 Doors Down,Away From The Sun,2002-01-01,6,54,False,3.97
6716,4OSBTYWVwsQhGLF9NHvIbR,Rolling in the Deep,Adele,21,2011-01-24,1,79,False,3.80
6720,1CkvWZme3pRgbzaxZnTl5X,Rolling in the Deep,Adele,21,2011-01-19,1,4,False,3.80
6718,3di5hcvxxciiqwMH1jarhY,Set Fire to the Rain,Adele,21,2011-01-24,5,82,False,4.04
6721,3CKCZ9pfwAfoMZlMncA1Nc,Set Fire to the Rain,Adele,21,2011-01-19,5,2,False,4.02
6719,3bNv3VuUOKgrf5hu3YcuRo,Someone Like You,Adele,21,2011-01-24,11,81,False,4.75
6722,4kflIGfjdZJW4ot2ioixTB,Someone Like You,Adele,21,2011-01-19,11,2,False,4.75
5011,4sPmO7WMQUAf45kwMOtONw,Hello,Adele,25,2016-06-24,1,0,False,4.92
5229,62PaSfnXSMyLshYJrlTuL3,Hello,Adele,25,2015-11-20,1,72,False,4.92


<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 10. Interpretação dos possíveis duplicados textuais</strong></font>

<font size="2" color='#66666'>

>- Registrar a interpretação técnica dos registros identificados como possíveis duplicidades textuais na base `spotify_data_clean.csv`.

>- **Resultado observado**
    - A análise identificou registros repetidos considerando a combinação:
        - `track_name + artist_name + album_name`
    - No entanto, esses registros possuem `track_id` diferentes, o que indica que não se tratam de duplicidades técnicas diretas.
    - A inspeção visual mostrou casos em que a mesma música aparece mais de uma vez com:
        - mesmo nome de faixa;
        - mesmo artista;
        - mesmo álbum;
        - mesmo número de faixa;
        - duração igual ou muito próxima;
        - datas de lançamento diferentes;
        - popularidades muito diferentes;
        - identificadores `track_id` distintos.

>- **Interpretação técnica**
    - Esses registros parecem representar múltiplas versões catalogadas da mesma música dentro do ecossistema Spotify.
        - Possíveis explicações incluem:
            - relançamentos;
            - edições de álbum;
            - versões regionais;
            - versões explícitas ou limpas;
            - reprocessamentos de catálogo;
            - duplicações provenientes da distribuição musical;
            - diferenças de popularidade entre versões antigas e atuais.
    - Portanto, não é adequado remover esses registros automaticamente nesta etapa.

>- **Risco analítico**
    - Apesar de não serem duplicidades técnicas por `track_id`, esses registros podem gerar distorções em algumas análises.
    - Possíveis impactos:

>| Análise | Risco |
>|---|---|
>| Ranking de músicas | A mesma música pode aparecer mais de uma vez |
>| Contagem de faixas por artista | O catálogo pode parecer maior do que é em termos canônicos |
>| Média de popularidade | Versões com popularidade zero ou muito baixa podem distorcer resultados |
>| Análise temporal | Datas alternativas podem criar ruído em séries históricas |
>| Análise por álbum | O mesmo álbum pode conter versões repetidas de faixas |

>- **Decisão metodológica**
    - A decisão nesta etapa é manter todos os registros na base principal, pois cada linha possui um `track_id` único.
    - No entanto, será criada uma marcação auxiliar para identificar possíveis duplicidades textuais.
    - A deduplicação definitiva, caso necessária, será tratada no Notebook 02 — Data Quality and Preparation.

>- **Estratégia futura**
    - No Notebook 02, poderão ser criadas duas visões da base:

>| Visão | Descrição |
>|---|---|
>| `df_tracks_catalog` | Mantém uma linha por `track_id`, preservando o catálogo Spotify |
>| `df_tracks_canonical` | Mantém uma linha por combinação canônica de música, artista e álbum |

>- Para a visão canônica, uma regra possível será manter o registro com maior `track_popularity` dentro de cada grupo textual.

>- **Resultado esperado**
    - A base continuará íntegra para o Data Understanding, mas os possíveis duplicados textuais serão documentados como ponto de atenção para a preparação dos dados.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Resumo dos grupos de possíveis duplicidades textuais</strong></font>

<font size="2" color='#66666'></font></div>

In [66]:
duplicate_key = ["track_name", "artist_name", "album_name"]

text_duplicate_groups = (
    df_main
    .groupby(duplicate_key, dropna=False)
    .agg(
        group_size=("track_id", "count"),
        unique_track_ids=("track_id", "nunique"),
        min_release_date=("album_release_date_parsed", "min"),
        max_release_date=("album_release_date_parsed", "max"),
        min_popularity=("track_popularity", "min"),
        max_popularity=("track_popularity", "max"),
        avg_popularity=("track_popularity", "mean"),
        min_duration=("track_duration_min", "min"),
        max_duration=("track_duration_min", "max"),
    )
    .reset_index()
)

text_duplicate_groups = text_duplicate_groups[text_duplicate_groups["group_size"] > 1].copy()

text_duplicate_groups["popularity_range"] = (
    text_duplicate_groups["max_popularity"] - text_duplicate_groups["min_popularity"]
)

text_duplicate_groups["duration_range"] = (
    text_duplicate_groups["max_duration"] - text_duplicate_groups["min_duration"]
)

text_duplicate_groups.sort_values(
    ["group_size", "popularity_range"],
    ascending=[False, False]
).head(30)

,track_name,artist_name,album_name,group_size,unique_track_ids,min_release_date,max_release_date,min_popularity,max_popularity,avg_popularity,min_duration,max_duration,popularity_range,duration_range
5642,S&M,Rihanna,Loud,3,3,2010-01-01,2010-11-16,55,84,70.33,4.05,4.06,29,0.01
7534,What's My Name?,Rihanna,Loud,3,3,2010-01-01,2010-11-16,49,76,64.00,4.38,4.38,27,0.00
6188,Sports car,Tate McRae,So Close To What,3,3,2025-02-20,2025-02-24,66,86,73.67,2.75,2.75,20,0.00
3303,I'm Not The Only One,Sam Smith,In The Lonely Hour,2,2,2014-01-01,2014-05-26,0,85,42.50,3.98,3.98,85,0.00
1010,Bring Me To Life,Evanescence,Fallen,2,2,2003-01-01,2003-03-04,3,87,45.00,3.93,3.93,84,0.00
1387,Complicated,Avril Lavigne,Let Go,2,2,2002-06-04,2002-08-26,0,83,41.50,4.07,4.07,83,0.00
6249,Stay With Me,Sam Smith,In The Lonely Hour,2,2,2014-01-01,2014-05-26,0,82,41.00,2.87,2.87,82,0.00
6085,Somebody That I Used To Know,Gotye,Making Mirrors,2,2,2011-01-01,2011-08-19,0,81,40.50,4.08,4.09,81,0.01
230,After Hours,The Weeknd,After Hours,2,2,2020-02-19,2020-03-20,1,81,41.00,6.01,6.03,80,0.02
1274,Chop Suey!,System Of A Down,Toxicity,2,2,2001-01-01,2001-09-04,5,85,45.00,3.50,3.50,80,0.00


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Grupos duplicados com maior diferença de popularidade</strong></font>

<font size="2" color='#66666'></font></div>

In [67]:
text_duplicate_groups.sort_values(
    "popularity_range",
    ascending=False
)[
    [
        "track_name",
        "artist_name",
        "album_name",
        "group_size",
        "min_release_date",
        "max_release_date",
        "min_popularity",
        "max_popularity",
        "popularity_range",
        "min_duration",
        "max_duration",
        "duration_range",
    ]
].head(30)

,track_name,artist_name,album_name,group_size,min_release_date,max_release_date,min_popularity,max_popularity,popularity_range,min_duration,max_duration,duration_range
3303,I'm Not The Only One,Sam Smith,In The Lonely Hour,2,2014-01-01,2014-05-26,0,85,85,3.98,3.98,0.00
1010,Bring Me To Life,Evanescence,Fallen,2,2003-01-01,2003-03-04,3,87,84,3.93,3.93,0.00
1387,Complicated,Avril Lavigne,Let Go,2,2002-06-04,2002-08-26,0,83,83,4.07,4.07,0.00
6249,Stay With Me,Sam Smith,In The Lonely Hour,2,2014-01-01,2014-05-26,0,82,82,2.87,2.87,0.00
6085,Somebody That I Used To Know,Gotye,Making Mirrors,2,2011-01-01,2011-08-19,0,81,81,4.08,4.09,0.01
1274,Chop Suey!,System Of A Down,Toxicity,2,2001-01-01,2001-09-04,5,85,80,3.50,3.50,0.00
3013,Hotline Bling,Drake,Views,2,2016-05-06,2016-05-13,0,80,80,4.45,4.45,0.00
5812,Set Fire to the Rain,Adele,21,2,2011-01-19,2011-01-24,2,82,80,4.02,4.04,0.02
1570,Dangerous Woman,Ariana Grande,Dangerous Woman,2,2016-04-01,2016-05-20,0,80,80,3.93,3.93,0.00
230,After Hours,The Weeknd,After Hours,2,2020-02-19,2020-03-20,1,81,80,6.01,6.03,0.02


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Salvamento de artefatos de diagnóstico de duplicidades textuais</strong></font>

<font size="2" color='#66666'></font></div>

In [68]:
TABLES_DIR = REPORTS_DIR / "tables"
TABLES_DIR.mkdir(parents=True, exist_ok=True)

# Artefato 1 — Detalhe dos registros possivelmente duplicados
possible_text_duplicates_export = possible_text_duplicates[
    [
        "track_id",
        "track_name",
        "artist_name",
        "album_name",
        "album_release_date",
        "track_number",
        "track_popularity",
        "explicit",
        "track_duration_min",
    ]
].copy()

# Artefato 2 — Resumo por grupo textual duplicado
text_duplicate_groups_export = text_duplicate_groups.sort_values(
    ["group_size", "popularity_range"],
    ascending=[False, False],
).copy()

# Caminhos de saída
possible_duplicates_path = TABLES_DIR / "possible_text_duplicates_detail.csv"
duplicate_groups_path = TABLES_DIR / "text_duplicate_groups_summary.csv"

# Salvamento em CSV
# sep=';' facilita abertura no Excel em ambiente PT-BR
possible_text_duplicates_export.to_csv(
    possible_duplicates_path,
    index=False,
    sep=";",
    encoding="utf-8-sig",
)

text_duplicate_groups_export.to_csv(
    duplicate_groups_path,
    index=False,
    sep=";",
    encoding="utf-8-sig",
)

print("Artefatos salvos com sucesso:")
print(f"- {possible_duplicates_path}")
print(f"- {duplicate_groups_path}")

Artefatos salvos com sucesso:
- D:\_DS-Projects\Data-Science\spotify-global-music-intelligence\reports\tables\possible_text_duplicates_detail.csv
- D:\_DS-Projects\Data-Science\spotify-global-music-intelligence\reports\tables\text_duplicate_groups_summary.csv


<div style="background-color:#f3f2ee">
    
<font size="3" color='#CC403E'><strong>✔️ Validação dos artefatos salvos</strong></font>

<font size="2" color='#66666'></font></div>

In [69]:
saved_artifacts_check = pd.DataFrame(
    {
        "artifact": [
            "possible_text_duplicates_detail.csv",
            "text_duplicate_groups_summary.csv",
        ],
        "path": [
            possible_duplicates_path,
            duplicate_groups_path,
        ],
        "exists": [
            possible_duplicates_path.exists(),
            duplicate_groups_path.exists(),
        ],
        "file_size_kb": [
            round(possible_duplicates_path.stat().st_size / 1024, 2)
            if possible_duplicates_path.exists()
            else None,
            round(duplicate_groups_path.stat().st_size / 1024, 2)
            if duplicate_groups_path.exists()
            else None,
        ],
    }
)

saved_artifacts_check

,artifact,path,exists,file_size_kb
0,possible_text_duplicates_detail.csv,D:\_DS-Projects\Data-Science\spotify-global-music-intelligence\reports\tables\possible_text_duplicates_detail.csv,True,20.82
1,text_duplicate_groups_summary.csv,D:\_DS-Projects\Data-Science\spotify-global-music-intelligence\reports\tables\text_duplicate_groups_summary.csv,True,10.95


<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 11. Salvamento dos artefatos de diagnóstico</strong></font>

<font size="2" color='#66666'>

>- Salvar os artefatos gerados durante a análise de possíveis duplicidades textuais da base `spotify_data_clean.csv`.

>- **Artefatos gerados**
    - Foram gerados dois arquivos CSV na pasta `reports/tables/`:

>| Artefato | Descrição |
>|---|---|
>| `possible_text_duplicates_detail.csv` | Lista detalhada dos registros com possível duplicidade textual |
>| `text_duplicate_groups_summary.csv` | Resumo agrupado por música, artista e álbum, com diferenças de popularidade, datas e duração |

>- **Justificativa técnica**
    - Esses arquivos preservam evidências da etapa de Data Understanding e serão usados como apoio para o Notebook 02 — Data Quality and Preparation.
    - A decisão atual é não remover automaticamente esses registros, pois cada linha possui um `track_id` único.
    - No entanto, os artefatos salvos permitem revisar os casos e definir uma regra futura para criação de uma visão canônica da base.

>- **Decisão**
    - Os registros serão mantidos na base principal nesta etapa.
    - A deduplicação canônica, se necessária, será tratada no Notebook 02, possivelmente criando duas visões:

>| Visão | Descrição |
>|---|---|
>| `df_tracks_catalog` | Mantém uma linha por `track_id` |
>| `df_tracks_canonical` | Mantém uma linha por música, artista e álbum, priorizando maior popularidade |

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 12. Síntese do Data Understanding</strong></font>

<font size="2" color='#66666'>

>- Consolidar os principais achados da fase de Data Understanding do projeto **Spotify Global Music Intelligence**.

>- Esta etapa resume o que foi aprendido sobre a base principal `spotify_data_clean.csv`, destacando estrutura, qualidade inicial,
    - cobertura temporal, duplicidades, limitações e decisões metodológicas para continuidade do projeto.

>- **Base principal definida**
    - Após avaliação inicial das fontes disponíveis, foi tomada a decisão metodológica de conduzir a EDA principal do Cap01 com o dataset:
        - `spotify_data_clean.csv`

>- O dataset complementar `spotify-tracks-dataset.csv` será preservado para uma etapa futura, possivelmente voltada a enriquecimento, ABT, análise de atributos musicais, segmentação ou modelagem.

>- Essa decisão mantém o projeto focado e evita complexidade prematura.

>- **Principais achados técnicos**

>| Dimensão | Achado | Interpretação |
>|---|---|---|
>| Volume | A base possui 8.582 registros e 15 colunas | Volume adequado para EDA inicial |
>| Chave técnica | `track_id` possui 8.582 valores únicos | Cada linha possui identificador técnico único |
>| Duplicidade exata | Não foram encontradas linhas totalmente duplicadas | Boa integridade inicial |
>| Duplicidade textual | Foram identificados 117 possíveis duplicados por música, artista e álbum | Prováveis versões, relançamentos ou registros alternativos |
>| Datas | `album_release_date` foi convertida sem erros | A base permite análise temporal |
>| Cobertura temporal | A base cobre registros entre 1952 e 2025 | Há faixas históricas além do recorte moderno |
>| Gêneros | `artist_genres` possui 39,16% de valores ausentes | Análises por gênero exigem cautela |
>| Artistas | `artist_name` possui apenas 3 valores ausentes | Problema pequeno, mas deve ser tratado no Notebook 02 |
>| Popularidade | `track_popularity` não possui valores ausentes | Variável central está disponível para EDA |

>- **Achado → Evidência → Implicação → Próxima ação**

>| Achado | Evidência | Implicação | Próxima ação |
>|---|---|---|---|
>| A base tem chave técnica consistente | `track_id` é 100% único | Podemos usar `track_id` como unidade técnica principal | Manter visão de catálogo por `track_id` |
>| Existem possíveis duplicados textuais | 117 casos por `track_name + artist_name + album_name` | Rankings podem ser distorcidos se a mesma música aparecer em múltiplas versões | Criar visão canônica no Notebook 02 |
>| A base possui registros anteriores a 2009 | Ano mínimo identificado: 1952 | O dataset contém catálogo histórico, não apenas lançamentos modernos | Criar variável de período analítico |
>| `artist_genres` tem alta ausência | 39,16% de nulos | Análises por gênero não devem ser interpretadas como completas | Tratar como `sem_genero_informado` |
>| Dataset complementar tem match parcial | Apenas 16,37% da base principal possui match por `track_id` | Não é adequado enriquecer a base principal nesta fase | Guardar para etapa futura |

>- **Decisões metodológicas tomadas**

>| Decisão | Justificativa |
>|---|---|
>| Usar `spotify_data_clean.csv` como base principal | Mantém o projeto focado no escopo inicial |
>| Não enriquecer com `spotify-tracks-dataset.csv` nesta fase | A interseção por `track_id` foi parcial |
>| Manter todos os registros por enquanto | Não há duplicidade técnica por `track_id` |
>| Investigar duplicados textuais no Notebook 02 | Podem representar versões reais do catálogo |
>| Criar variável temporal auxiliar | A base contém registros de 1952 a 2025 |
>| Tratar ausência de gênero como limitação | `artist_genres` tem percentual relevante de nulos |
>| Salvar artefatos de duplicidade | Garante rastreabilidade e apoio ao tratamento futuro |

>- **Limitações identificadas**

>- A fase de Data Understanding também revelou algumas limitações importantes:
    - o dataset contém registros anteriores ao recorte inicialmente esperado de 2009–2025;
    - a variável `artist_genres` possui alta taxa de ausência;
    - algumas músicas aparecem em múltiplas versões textualmente semelhantes;
    - a popularidade pode refletir um snapshot temporal específico;
    - não há atributos musicais detalhados, como energia, dançabilidade ou valência, nesta base principal;
    - o dataset complementar possui apenas interseção parcial com a base principal.

>- Essas limitações não impedem a continuidade do projeto, mas deverão ser consideradas na preparação dos dados e na interpretação da EDA.

>- **Conclusão da etapa**
    - A base `spotify_data_clean.csv` foi considerada adequada para continuidade do projeto.
    - Ela possui boa integridade estrutural, chave técnica consistente, variável de popularidade completa e cobertura temporal ampla.
    - Os principais pontos que exigem tratamento no próximo notebook são:
        1. conversão e padronização de datas;
        2. criação de variáveis temporais;
        3. tratamento de `artist_genres` ausente;
        4. investigação dos registros sem `artist_name`;
        5. marcação ou criação de visão canônica para possíveis duplicados textuais;
        6. preparação de uma base Silver confiável para EDA.

---
</font></div>

<div style="background-color:#f3f2ee">
    
<font size="4" color='#CC403E'><strong>📌 13. Critérios de passagem para o Notebook 02</strong></font>

<font size="2" color='#66666'>

>- Definir os critérios mínimos para avançar do **Notebook 01 — Data Understanding** para o **Notebook 02 — Data Quality and Preparation**.

>- O objetivo é garantir que a próxima fase comece com clareza sobre o que precisa ser tratado, padronizado e preparado.

>- **Critérios avaliados**

>| Critério | Status | Observação |
>|---|---|---|
>| Base principal definida | Concluído | `spotify_data_clean.csv` será a fonte principal |
>| Arquivo carregado corretamente | Concluído | Dataset carregado sem erros críticos |
>| Estrutura inicial compreendida | Concluído | 8.582 registros e 15 colunas |
>| Chave técnica avaliada | Concluído | `track_id` é único |
>| Duplicidade exata avaliada | Concluído | Nenhuma linha totalmente duplicada |
>| Duplicidade textual avaliada | Concluído | 117 possíveis casos identificados |
>| Artefatos de duplicidade salvos | Concluído | CSVs salvos em `reports/tables/` |
>| Tipos de dados avaliados | Concluído | Algumas conversões serão necessárias |
>| Datas avaliadas | Concluído | Datas válidas entre 1952 e 2025 |
>| Valores ausentes avaliados | Concluído | Atenção para `artist_genres` e `artist_name` |
>| Fonte complementar analisada | Concluído | Uso adiado para etapa futura |
>| Próximas ações definidas | Concluído | Tratamentos serão feitos no Notebook 02 |

>- **Decisão de passagem**
    - A fase de Data Understanding está suficientemente concluída para avançar ao **Notebook 02 — Data Quality and Preparation**.
    - A próxima etapa terá como foco transformar o diagnóstico obtido neste notebook em ações práticas de preparação dos dados.

>- **Ações previstas para o Notebook 02**
    - O Notebook 02 deverá executar:
        1. criação de cópia controlada da base principal;
        2. padronização de tipos;
        3. conversão de `album_release_date` para data;
        4. criação de `release_year`, `release_month` e `periodo_analitico`;
        5. tratamento de `artist_genres` ausente;
        6. investigação dos registros sem `artist_name`;
        7. criação de marcação para possíveis duplicados textuais;
        8. construção de visão de catálogo por `track_id`;
        9. possível criação de visão canônica por música, artista e álbum;
        10. salvamento da base preparada na camada Silver;
        11. geração de relatório de qualidade dos dados.

>- **Encerramento do Notebook 01**
    - O **Notebook 01 — Data Understanding** cumpriu seu papel de compreender a base principal antes da preparação e da análise exploratória.
    - A partir deste ponto, o projeto está pronto para avançar para a fase de **Data Quality and Preparation**,
        - onde os problemas identificados serão tratados de forma controlada, documentada e reprodutível.

---
</font></div>

In [70]:
%reload_ext watermark 
%watermark -a "Roberto-SSoares-LfLngLrnng" -d -t -u --iversions -v -m -h

Author: Roberto-SSoares-LfLngLrnng

Last updated: 2026-05-05 03:19:49

Python implementation: CPython
Python version       : 3.12.12
IPython version      : 9.13.0

Compiler    : MSC v.1944 64 bit (AMD64)
OS          : Windows
Release     : 11
Machine     : AMD64
Processor   : Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
CPU cores   : 4
Architecture: 64bit

Hostname: PC-ROBERTO

IPython: 9.13.0
numpy  : 2.4.4
pandas : 3.0.2



<div style="background-color:#f3f2ee">
    
<font size="6" color='#CC403E'><strong>FIM</strong></font>

<font size="2" color='#66666'></font></div>